components/eam/src/physics/cam/dycoms_rad.F90

```diff
@@ -44,60 +61,70 @@ subroutine dycoms_radiation_tend(state, ptend, pbuf, ztodt)
                           F1 = 22.0_r8, &
                           kappa = 85.0_r8
 
-   real(r8), pointer, dimension(:,:) :: iclwp, cld
-
    integer :: i, k
 
+   !-----------------------------------------------------------------------------------
    ncol = state%ncol
    lchnk = state%lchnk
-   pint(1:ncol,1:pver+1) = state%pint(1:ncol,1:pver+1)
 
-   call pbuf_get_field(pbuf, pbuf_get_index('ICLWP'), iclwp)
-   call pbuf_get_field(pbuf, pbuf_get_index('CLD'), cld)
+   ! Calculate liquid water path integrated from model top to a layer interface of interest.
+   ! Note that in this highly idealized case, we are not making any assumptions about cloud overlap,
+   ! hence grid-box mean cloud liquid mixing ratio is used here.
 
-   lwp = 0.0_r8
-   do k=2,pver+1
-      do i = 1,ncol
-         lwp(i,k) = lwp(i,k-1) + iclwp(i,k-1)*cld(i,k-1)
-      end do
+   call cnst_get_ind('CLDLIQ',ixcldliq)
+
+   lwp_abv(:,:) = 0.0_r8
+   do k=2,pverp
+      lwp_in_layer(:ncol) = state%q(:ncol,k-1,ixcldliq) * state%pdel(:ncol,k-1) /gravit
+      lwp_abv(:ncol,k) = lwp_abv(:ncol,k-1) + lwp_in_layer(:ncol) 
    end do
 
-   do k = 1, pver+1
+   ! Calculate upwelling LW flux at layer interfaces
+   ! using the first 2 RHS terms in Eq. (3) of Stevens et al. (2005), "Evaluation of
+   ! Large-Eddy Simulations via Observations of Nocturnal Marine Stratocumulus"
+
+   do k = 1, pverp
       do i = 1,ncol
-         radflux(i, k) = F0*exp(-kappa*lwp(i,k)) + F1*exp(-kappa*(lwp(i,pver+1)-lwp(i,k)))
+         radflux(i, k) = F0*exp(-kappa*lwp_abv(i,k)) &
+                       + F1*exp(-kappa*(lwp_abv(i,pverp)-lwp_abv(i,k)))
       end do
    end do
 
+   ! Calculate cp*dT/dt
+
    do k = 1,pver
-      do i = 1,ncol
-         qrl(i,k) = (radflux(i,k+1)-radflux(i,k))*gravit/(pint(i,k+1)-pint(i,k))
-      end do
+      qrl(:ncol,k) = (radflux(:ncol,k+1)-radflux(:ncol,k)) * gravit / state%pdel(:ncol,k)
    end do
 
+   !--------------------------------------------------------------------------
+   ! Assign values to arrays needed by the host model
+   !--------------------------------------------------------------------------
    call physics_ptend_init(ptend, state%psetcols, 'dycoms_radheat', ls=.true.) 
    ptend%s(:ncol,:) = qrl(:ncol,:)
 
+   ! Net flux into the atm column = net incoming LW at sfc - net outgoing LW at TOP.
```

Comment:
Maybe clarify this comment by saying '... = net LW at sfc incoming to the atm column - net outgoing LW at TOP.'?